File: M68000/Tests/M68000-Generated/python/generator/types/test_data.py
"""
This module contains the data structures that represent a test case.
"""

import dataclasses
from typing import Dict, List

from .status_register import StatusRegister

Registers = Dict[int, int]
Memory = Dict[int, int]


# pylint: disable=invalid-name
@dataclasses.dataclass(kw_only=True)
class GivenConditions:
    status_register: StatusRegister
    pc: int
    d: Registers = dataclasses.field(default_factory=dict)
    a: Registers = dataclasses.field(default_factory=dict)
    memory: Memory = dataclasses.field(default_factory=dict)
    usp: int = 0
    ssp: int = 0
    stopped: bool = False


# pylint: disable=invalid-name
@dataclasses.dataclass(kw_only=True)
class ThenConditions:
    status_register: StatusRegister
    pc: int
    cycles: int
    d: Registers = dataclasses.field(default_factory=dict)
    a: Registers = dataclasses.field(default_factory=dict)
    memory: Memory = dataclasses.field(default_factory=dict)
    usp: int = 0
    ssp: int = 0
    stopped: bool = False


@dataclasses.dataclass(kw_only=True)
class TestData:
    given: GivenConditions
    when: List[str]
    then: ThenConditions

File: M68000/Tests/M68000-Generated/src/TestData.cpp
#include <nlohmann/json.hpp>
#include <fstream>
#include "TestData.h"

void from_json(const nlohmann::json& j, StatusRegister& statusRegister) {
    statusRegister.t = j.at("t");
    statusRegister.s = j.at("s");
    statusRegister.interrupt_mask = j.at("interrupt_mask");
    statusRegister.x = j.at("x");
    statusRegister.n = j.at("n");
    statusRegister.z = j.at("z");
    statusRegister.v = j.at("v");
    statusRegister.c = j.at("c");
}

void from_json(const nlohmann::json& j, Given& given) {
    given.status_register = j.at("status_register");
    given.pc = j.at("pc");
    given.d = j.at("d");
    given.a = j.at("a");
    given.memory = j.at("memory");
    given.usp = j.at("usp");
    given.ssp = j.at("ssp");
    given.stopped = j.at("stopped");
}

void from_json(const nlohmann::json& j, Then& then) {
    then.status_register = j.at("status_register");
    then.pc = j.at("pc");
    then.d = j.at("d");
    then.a = j.at("a");
    then.memory = j.at("memory");
    then.cycles = j.at("cycles");
    then.usp = j.at("usp");
    then.ssp = j.at("ssp");
    then.stopped = j.at("stopped");
}

void from_json(const nlohmann::json& j, TestData& data) {
    data.given = j.at("given");
    data.when = j.at("when");
    data.then = j.at("then");
}

std::vector<TestData> GetTests(const std::string& path) {
    std::ifstream input(path);
    return nlohmann::json::parse(input);
}